hera_cal/vis_clean.py

```diff
@@ -7,29 +7,240 @@
-from . import apply_cal
-from . import version
+from . import io, apply_cal, version, redcal
 from .datacontainer import DataContainer
 from .utils import echo
 from .flag_utils import factorize_flags
-import warnings
+
+
+def truncate_flagged_edges(data_in, weights_in, x, ax='freq'):
+    """
+    cut away edge channels and integrations that are completely flagged
+
+    Parameters
+    ----------
+    data_in : array-like, 2d
+        data from which to remove edge integrations and channels that are completely flagged.
+    weights_in : array-like, 2d
+        weights to determine which edge integrations and channels are completely flagged
+        (close to zero). Will also be truncated.
+    x : array-like, 1-d (or 2-tuple/list of arrays)
+        x-values for data axis that we are truncating.
+        if ax=='both', should be a 2-tuple or 2-list of 1-d arrays.
+    ax : string, optional
+        axis to truncate flagged edges from. Should be 'freq' to truncate
+        completely flagged edge channels, 'time' to truncate completely flagged
+        edge integrations, or 'both' to do both.
+
+    Returns
+    -------
+    xout : array-like 1d (or 2-list if ax=='both').
+            x with completely flagged edges trimmed off.
+    dout : array-like 2d.
+            data_in with completely flagged edges trimmed off.
+    wout : array-like 2d.
+            weights_in with completely flagged edges trimmed off.
+    edges : 2-list of 2-tuples
+            the width of the edges trimmed.
+    """
+    # if axis == 'time', just use freq mode
+    # on transposed arrays.
+    if ax == 'time':
+        xout, dout, wout, edges = truncate_flagged_edges(data_in.T, weights_in.T, x)
+        dout = dout.T
+        wout = wout.T
+        edges = [edges[1], (0, 0)]
+    else:
+        ind_left = 0
+        ind_right = data_in.shape[1]
+        # Identify edge channels that are flagged.
+        unflagged_chans = np.where(~np.all(np.isclose(weights_in, 0.0), axis=0))[0]
+        if np.count_nonzero(unflagged_chans) > 0:
+            # truncate data to be filtered where appropriate.
+            ind_left = np.min(unflagged_chans)
+            ind_right = np.max(unflagged_chans) + 1
+        dout = data_in[:, ind_left: ind_right]
+        wout = weights_in[:, ind_left: ind_right]
+        edges = (ind_left, data_in.shape[1] - ind_right)
+        if ax == 'both':
+            x1 = x[1][ind_left: ind_right]
+            x0, dout, wout, e0 = truncate_flagged_edges(dout, wout, x[0], ax='time')
+            xout = [x0, x1]
+            edges = [e0[0], edges]
+        else:
+            xout = x[ind_left: ind_right]
+            edges = [(0, 0), edges]
+    return xout, dout, wout, edges
+
+
+def flag_rows_with_flags_within_edge_distance(weights_in, min_flag_edge_distance, ax='freq'):
+    """
+    flag integrations (and/or channels) with flags within min_flag_edge_distance of edge.
+
+    Parameters
+    -----------
+    weights_in : array-like, 2d
```

Comment:
same comment here about shape of input data

Reply:
added this.